### DISCUSSION TITLE
Upgrading Toit to v2.0

### DISCUSSION THREAD

**Discussion by askdesigners** [toitlang/toit#1139] [General]
I've had Toit installed for about a year now, and have been dabbling with on and off. I recently picked it up again, and found that a few examples from the docs don't work and discovered I'm using v1 still. 

Is there an upgrade path?

**Comment by floitsch** [comment]
There are a few things that have changed between v1 and v2.

## Pubsub
The biggest one, is the absence of pubsub. Toit v1 had a pubsub mechanism to send data from/to the device. In v2 we recommend to use MQTT instead. For small projects io.adafruit.com might be suitable, but there are many others.

## BLE
Similarly, the BLE library was changed. Fortunately, the v1 implementation was relatively limited, and upgrading to v2 should be relatively straight-forward. If you encounter challenges just contact us on our [Discord chat](https://discord.gg/Q7Y9VQ5nh2).

## Specification files
In v1, deploying an application consisted of writing a specification file and then installing the application through that file. The specification file informed the system when it should execute the application.

When no application was running and no connection to the server was needed, the system would go into deep sleep, saving power.

In v2, with Jaguar, an application is installed with `jag container install name entry.toit`. However, this installation comes without any specification file, and the system just runs the application whenever it is booted.

Here are some ways to get similar features to v1:
* Import the `esp32` library and use `deep_sleep` to put the device into deep sleep. The function takes a duration as argument.
* Use `sleep`.
* Use `enable_touchpad_wakeup` and `enable_external_wakeup` from the `esp32` library to wake the system from deep sleep.

## Others?
If you see other changes, let me know, and I will try to update this post.

**Reply by askdesigners** [reply]
I'm more wondering about the upgrade process from v1 -> v2

I deleted the `toit` binary from my applications folder, installed jag with homebrew, but now the `toit` command doesn't work any more. Is that ok i suppose?

**Reply by askdesigners** [reply]
![Screen Shot 2022-10-21 at 13 06 23](https://user-images.githubusercontent.com/2891457/197182015-84bb894d-ea83-474d-90d8-5c4454ceb00e.png)

**Reply by floitsch** [reply]
The easiest is to uninstall toit, and to install [Jaguar](https://github.com/toitlang/jaguar) instead.

And yes: that removes the `toit` command. You should just use `jag` instead.

If the vscode extension is complaining, have a look at the extension settings. There might be one that still refers to the `toit` command. Usually, you shouldn't need any path-related entry for the extension.

**Reply by askdesigners** [reply]
ok great thanks. I might suggest using a major version number bump when there are breaking functionality changes on the extension. I had one patch version lower than the latest and didn't realize there was a breaking change from one to the other. 

Also, I'm kind of sad about losing that spec file! It was a really neat feature of the language! Was it not getting used by many people? I loved the ability to spin up a task on a cron without writing any procedural code at all.

**Reply by floitsch** [reply]
The extension is supposed to continue to work, but having both `toit` and `jag` reachable confuses it a bit.

We will try to bring the specification file back.

**Reply by askdesigners** [reply]
Appreciate the support! Toit is really amazing for IoT dev.

**Comment by micrcx** [comment]
It should be understood that version v2 works exclusively with **Jaguar** and it's not possible to use the **toit** console command to update the sdk further? Anyway, I tried make this:
```
michael_k@michaelk:~/toit_apps/Hsm3/tests$ toit sdk update
An SDK with version 'v1.6.24' is already installed locally. Do you want to override? [y/n] ^C
```
and I see that it's impossible to update toit via console to version v2. This is "not yet possible", ie. something temporary, or is it from now and forever: need to switch to using **jag** instead of **toit**? Or is there still some way to update the SDK to v2 and continue use **toit**? 
PS. Personally, I was more than satisfied with the way of working with the **toit** console command, because it allowed to run apps both on the computer and on the chip, update the SDK & much more...

**Reply by floitsch** [reply]
We are planning to provide a `toit sdk update` that migrates a device from Toit v1 to the v2. At that point the device wouldn't connect to the Toit console anymore, but would become a Jaguar device.

Jaguar, too, has an option to run locally: `jag run -d host hello.toit`

We are working on a product that will make it easier to manage devices and bringing some of the v1 functionality back. The top item here, is to be able to upgrade devices without them being in the same network. (Something our customers need). In some form, we will also have to deal with different versions of the SDK.

**Reply by ghost** [reply]
That is, if I understand correctly, updating the SDK from **v1.x** to **v2.x** will cause the **toit** command **to die** for devices updated to **v2**. The same fate awaits **pubsub**, your link **https://console.toit.io/**, and the **GRPC API**?

**Reply by floitsch** [reply]
Correct.

**Reply by ghost** [reply]
Well, how then will it be possible to send commands or data to **esp32** from a **phone** or **computer** and receive them back, providing some interaction over the internet? Again back to **tcp/ip** and **udp**?

**Reply by floitsch** [reply]
We recommend MQTT.

There are a few servers that are free for moderate use.
One of the easiest to set up is io.adafruit.com. It has a limit on how many messages can be sent per minute (30 iirc), but is easy to set up and allows for nice dashboards.

We have an example for how to use it here: https://github.com/toitware/mqtt/blob/main/examples/adafruit.toit

**Reply by ghost** [reply]
A couple more questions, if I may.
1) In version **v1**, there are ***.yaml** exists files that allow you to manage the life cycle of an application on a chip. For example, run it on triggers at a certain time or periodically, or upon receipt of a message via pubsub. If specification files,  pubsub and toit cloud will be canceled in version **v2**, similar functions should either be taken over by the **MQTT broker**, or these features will not be available at all. Ie. in the best case, in addition to chip programming, you will also have to do programming on the cloud, which will complicate development.
2) In version **v1**, pubsub also allowed messaging between applications **inside** esp32. Will some other data exchange mechanism be proposed?

**Reply by floitsch** [reply]
For (1), we are working on a new tool that should eventually have similar functionality. That said, starting applications based on MQTT messages might take a bit longer. It shouldn't be too hard, but it depends on our priorities.

For (2): the closest is probably "services". Individual applications can implement and expose a service that other applications can then connect to. See, for example, a temperature service here: https://github.com/kasperl/toit-sensors/blob/main/src/temperature.toit

We are still polishing the API, so there isn't that much documentation yet, but this mechanism could then be used to either communication directly between two apps, or to implement a "pubsub" service that takes messages and distributes them among the apps that have registered.
